compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -5961,6 +5961,14 @@ private void visitCheckAndCheckPanicExpr(BLangCheckedExpr checkedExpr, AnalyzerD
                         DiagnosticWarningCode.CHECKED_EXPR_INVALID_USAGE_NO_ERROR_TYPE_IN_RHS,
                         operatorType);
                 data.resultType = checkedExpr.expr.getBType();
+
+                BLangTypeConversionExpr impConversionExpr = checkedExpr.expr.impConversionExpr;
+                if (impConversionExpr != null) {
+                    // Exclude any manually added error type, by reassigning
+                    BType expType = Types.getReferredType(data.expType);
+                    impConversionExpr.setBType(expType);
```

Comment:
We do. Consider `T  _ = check expr`,

Here when we do the type checking for the `expr`, the contextually excepted type is `T|error`.
Only after the type-checking, we know whether the actual type includes an error type or not. If it does not include an error type, usage of checking keyword is redundant. i.e. `T  _ = check expr` is equivalent to `T _ = expr`. Because we did the type checking with expected type being `T|error` earlier, `impConversionExpr` is not correct. So we have reset  it with the type `T`.

Reply:
Thank you for the details. I only have a small concern about why we do this twice. Why can't we add the implicit conversion here or at desugar when we have all the details. Otherwise, all looks good.

Reply:
Had a discussion I will look at this logic more and reply.